test/test_swa.py

```diff
@@ -0,0 +1,692 @@
+
+    def _define_vars_loss_opt(self):
+        x = Variable(torch.Tensor([5., 2.]), requires_grad=True)
+        y = Variable(torch.Tensor([3., 7.]), requires_grad=True)
+
+        def loss_fun(a, b):
+            return torch.sum(a * b)**2
+
+        opt = optim.SGD([{'params': [x]},
+                        {'params': [y], 'lr': 1e-3}], lr=1e-2, momentum=0.9)
+        return x, y, loss_fun, opt
+
+    def testSWAAuto(self):
+        # Tests SWA in Auto mode: values of x and y after opt.swap_swa_sgd()
+        # should be equal to the manually computed averages
+        x, y, loss_fun, opt = self._define_vars_loss_opt()
+        swa_start = 5
+        swa_freq = 2
+        opt = contriboptim.SWA(opt, swa_start=swa_start, swa_freq=swa_freq, swa_lr=0.001)
+
+        x_sum = torch.zeros_like(x)
+        y_sum = torch.zeros_like(y)
+        n_avg = 0
+        for i in range(1, 11):
+            opt.zero_grad()
+            loss = loss_fun(x, y)
+            loss.backward()
+            opt.step()
+            if i % swa_freq == 0 and i > swa_start:
+                n_avg += 1
+                x_sum += x.data
+                y_sum += y.data
+
+        opt.swap_swa_sgd()
+        x_avg = x_sum / n_avg
+        y_avg = y_sum / n_avg
+        self.assertEqual(x_avg, x)
+        self.assertEqual(y_avg, y)      
+
+    def testSWAManual(self):
+        # Tests SWA in manual mode: values of x and y after opt.swap_swa_sgd()
+        # should be equal to the manually computed averages
+        x, y, loss_fun, opt = self._define_vars_loss_opt()
+        opt = contriboptim.SWA(opt)
+        swa_start = 5
+        swa_freq = 2
+
+        x_sum = torch.zeros_like(x)
+        y_sum = torch.zeros_like(y)
+        n_avg = 0
+        for i in range(1, 11):
+            opt.zero_grad()
+            loss = loss_fun(x, y)
+            loss.backward()
+            opt.step()
+            if i % swa_freq == 0 and i > swa_start:
+                opt.update_swa()
+                n_avg += 1
+                x_sum += x.data
+                y_sum += y.data
+
+        opt.swap_swa_sgd()
+        x_avg = x_sum / n_avg
+        y_avg = y_sum / n_avg
+        self.assertEqual(x_avg, x)
+        self.assertEqual(y_avg, y)      
+
+    def testSWAManualGroup(self):
+        # Tests SWA in manual mode with only y param group updated: 
+        # value of x should not change after opt.swap_swa_sgd() and y should
+        # be equal to the manually computed average
+        x, y, loss_fun, opt = self._define_vars_loss_opt()
+        opt = contriboptim.SWA(opt)
+        swa_start = 5
+        swa_freq = 2
+
+        y_sum = torch.zeros_like(y)
+        n_avg = 0
+        for i in range(1, 11):
+            opt.zero_grad()
+            loss = loss_fun(x, y)
+            loss.backward()
+            opt.step()
+            if i % swa_freq == 0 and i > swa_start:
+                opt.update_swa_group(opt.param_groups[1])
+                n_avg += 1
+                y_sum += y.data
```

Comment:
This code keeps repeating over and over in those tests. Is there really no way to factor out this training loop into a separate function? You can pass the updating line as a lambda to be called inside.

Reply:
Done, see https://github.com/izmailovpavel/contrib/blob/master/test/test_swa.py#L337-L343